File: services/academic-service/src/test/java/com/visor/school/academicservice/repository/TeacherAssignmentRepositoryTest.java
package com.visor.school.academicservice.repository;

import com.visor.school.academicservice.model.TeacherAssignment;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.test.context.ActiveProfiles;

import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

@DataJpaTest
@ActiveProfiles("test")
class TeacherAssignmentRepositoryTest {

    @Autowired
    private TestEntityManager entityManager;

    @Autowired
    private TeacherAssignmentRepository teacherAssignmentRepository;

    private TeacherAssignment testAssignment;
    private UUID testTeacherId;
    private UUID testClassId;

    @BeforeEach
    void setup() {
        testTeacherId = UUID.randomUUID();
        testClassId = UUID.randomUUID();
        testAssignment = new TeacherAssignment(testTeacherId, testClassId, false);
        entityManager.persistAndFlush(testAssignment);
    }

    @Test
    void shouldFindAssignmentsByTeacherId() {
        List<TeacherAssignment> assignments = teacherAssignmentRepository.findByTeacherId(testTeacherId);
        assertFalse(assignments.isEmpty());
        assertTrue(assignments.stream().allMatch(a -> a.getTeacherId().equals(testTeacherId)));
    }

    @Test
    void shouldFindAssignmentsByClassId() {
        List<TeacherAssignment> assignments = teacherAssignmentRepository.findByClassId(testClassId);
        assertFalse(assignments.isEmpty());
        assertTrue(assignments.stream().allMatch(a -> a.getClassId().equals(testClassId)));
    }

    @Test
    void shouldFindClassTeacherAssignments() {
        TeacherAssignment classTeacherAssignment = new TeacherAssignment(UUID.randomUUID(), testClassId, true);
        entityManager.persistAndFlush(classTeacherAssignment);

        List<TeacherAssignment> classTeachers = teacherAssignmentRepository.findClassTeacherByClassId(testClassId);
        assertFalse(classTeachers.isEmpty());
        assertTrue(classTeachers.stream().allMatch(a -> a.isClassTeacher() && a.getClassId().equals(testClassId)));
    }

    @Test
    void shouldCheckIfAssignmentExists() {
        assertTrue(teacherAssignmentRepository.existsByTeacherIdAndClassId(testTeacherId, testClassId));
        assertFalse(teacherAssignmentRepository.existsByTeacherIdAndClassId(UUID.randomUUID(), testClassId));
    }

    @Test
    void shouldFindAssignmentByTeacherAndClass() {
        List<TeacherAssignment> assignments = teacherAssignmentRepository.findByTeacherIdAndClassId(testTeacherId, testClassId);
        assertFalse(assignments.isEmpty());
        assertEquals(testTeacherId, assignments.get(0).getTeacherId());
        assertEquals(testClassId, assignments.get(0).getClassId());
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/repository/TeacherAssignmentRepositoryTest.kt
package com.visor.school.academicservice.repository

import com.visor.school.academicservice.model.TeacherAssignment
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager
import org.springframework.test.context.ActiveProfiles
import java.util.UUID

@DataJpaTest
@ActiveProfiles("test")
class TeacherAssignmentRepositoryTest @Autowired constructor(
    private val entityManager: TestEntityManager,
    private val teacherAssignmentRepository: TeacherAssignmentRepository
) {

    private lateinit var testAssignment: TeacherAssignment
    private val testTeacherId = UUID.randomUUID()
    private val testClassId = UUID.randomUUID()

    @BeforeEach
    fun setup() {
        testAssignment = TeacherAssignment(
            teacherId = testTeacherId,
            classId = testClassId,
            isClassTeacher = false
        )
        entityManager.persistAndFlush(testAssignment)
    }

    @Test
    fun `should find assignments by teacherId`() {
        val assignments = teacherAssignmentRepository.findByTeacherId(testTeacherId)

        assertTrue(assignments.isNotEmpty())
        assertTrue(assignments.all { it.teacherId == testTeacherId })
    }

    @Test
    fun `should find assignments by classId`() {
        val assignments = teacherAssignmentRepository.findByClassId(testClassId)

        assertTrue(assignments.isNotEmpty())
        assertTrue(assignments.all { it.classId == testClassId })
    }

    @Test
    fun `should find class teacher assignments`() {
        val classTeacherAssignment = TeacherAssignment(
            teacherId = UUID.randomUUID(),
            classId = testClassId,
            isClassTeacher = true
        )
        entityManager.persistAndFlush(classTeacherAssignment)

        val classTeachers = teacherAssignmentRepository.findClassTeacherByClassId(testClassId)

        assertTrue(classTeachers.isNotEmpty())
        assertTrue(classTeachers.all { it.isClassTeacher && it.classId == testClassId })
    }

    @Test
    fun `should check if assignment exists`() {
        assertTrue(teacherAssignmentRepository.existsByTeacherIdAndClassId(testTeacherId, testClassId))
        assertFalse(teacherAssignmentRepository.existsByTeacherIdAndClassId(UUID.randomUUID(), testClassId))
    }

    @Test
    fun `should find assignment by teacher and class`() {
        val assignments = teacherAssignmentRepository.findByTeacherIdAndClassId(testTeacherId, testClassId)

        assertTrue(assignments.isNotEmpty())
        assertEquals(testTeacherId, assignments.first().teacherId)
        assertEquals(testClassId, assignments.first().classId)
    }
}

